allennlp/fairness/bias_mitigators.py

```diff
@@ -0,0 +1,542 @@
+            raise ConfigurationError("bias_direction1 and bias_direction2 must be one-dimensional.")
+        if evaluation_embeddings.size(-1) != bias_direction1.size(-1) or evaluation_embeddings.size(
+            -1
+        ) != bias_direction2.size(-1):
+            raise ConfigurationError(
+                "All embeddings, bias_direction1, and bias_direction2 must have the same dimensionality."
+            )
+        if bias_direction1.size(-1) < 2:
+            raise ConfigurationError(
+                "Dimensionality of all embeddings, bias_direction1, and bias_direction2 must \
+                be >= 2."
+            )
+
+        with torch.set_grad_enabled(self.requires_grad):
+            bias_direction1 = bias_direction1 / torch.linalg.norm(bias_direction1)
+            bias_direction2 = bias_direction2 / torch.linalg.norm(bias_direction2)
+
+            bias_direction2_orth = self._remove_component(
+                bias_direction2.reshape(1, -1), bias_direction1
+            )[0]
+            bias_direction2_orth = bias_direction2_orth / torch.linalg.norm(bias_direction2_orth)
+
+            # Create rotation matrix as orthonormal basis
+            # with v1 and v2'
+            init_orth_matrix = torch.eye(
+                bias_direction1.size(0),
+                device=evaluation_embeddings.device,
+                requires_grad=self.requires_grad,
+            )
+            rotation_matrix = torch.zeros(
+                (bias_direction1.size(0), bias_direction1.size(0)),
+                device=evaluation_embeddings.device,
+                requires_grad=self.requires_grad,
+            )
+            rotation_matrix = torch.cat(
+                [
+                    bias_direction1.reshape(1, -1),
+                    bias_direction2_orth.reshape(1, -1),
+                    rotation_matrix[2:],
+                ]
+            )
+            # Apply Gram-Schmidt
+            for i in range(len(rotation_matrix) - 2):
+                subspace_proj = torch.sum(
+                    self._proj(
+                        rotation_matrix[: i + 2].clone(), init_orth_matrix[i], normalize=True
+                    ),
+                    dim=0,
+                )
+                rotation_matrix[i + 2] = (init_orth_matrix[i] - subspace_proj) / torch.linalg.norm(
+                    init_orth_matrix[i] - subspace_proj
+                )
+
+            mask = ~(evaluation_embeddings == 0).all(dim=-1)
+            # Transform all evaluation embeddings
+            # using orthonormal basis computed above
+            rotated_evaluation_embeddings = torch.matmul(
+                evaluation_embeddings[mask], rotation_matrix.t()
+            )
+            # Want to adjust first 2 coordinates and leave d - 2
+            # other orthogonal components fixed
+            fixed_rotated_evaluation_embeddings = rotated_evaluation_embeddings[..., 2:]
+            # Restrict attention to subspace S
```

Comment:
where subspace S is ...?

Reply:
the subspace spanned by the bias directions (made a comment in the file)